upgrade_analysis/odoo_patch/odoo/addons/base/models/ir_model.py

```diff
@@ -32,11 +32,11 @@ def cons_text(txt):
             if record:
                 xml_id = f"{module}.constraint_{conname}"
                 data_list.append(dict(xml_id=xml_id, record=record))
-        if data_list:
-            self.env["ir.model.data"]._update_xmlids(data_list)
             # Begin OpenUpgrade addition
-            for data in data_list:
-                xml_id = data.get("xml_id")
-                module = xml_id.split(".")[0]
-                upgrade_log.log_xml_id(self.env.cr, module, xml_id)
+            xml_id = f"{module}.constraint_{conname}"
+            upgrade_log.log_xml_id(self.env.cr, module, xml_id)
+            self.pool.loaded_xmlids.add(xml_id)
```

Comment:
this one I found dubious in the beginning but it makes sense: We init the modules twice, after the first loading process `pool.loaded_xmlids` is cleared, and the second one won't see those xmlids because of the change in core linked above. Standard Odoo doesn't encounter this issue, so no problem for them, but we have to take care that the xmlids of constraints are touched because otherwise Odoo deletes all constraints in the second `_process_end`, leading to my intermediate result with the definitions being empty

Reply:
Good find! If I understand correctly, vanilla Odoo does not mark the xml id of a previously existing constraint as loaded. But why does vanilla Odoo not remove the reflected constraints after a module update like any other xml id that is not registered as loaded? Can you help me understand?

Reply:
oh I see now it does:

https://github.com/OCA/OCB/blob/17.0/odoo/addons/base/models/ir_model.py#L1882-L1883

that's not part of the diff I linked, and I missed this change when comparing the function to the current code. It's functionally equivalent to the above, but to be in line with the current code I refreshed the function from HEAD in a fixup commit so you see what's going on (let me know when to squash this).

While being on it I excluded the whole odoo patch folder from pre-commit, so that we stay as close as possible to the code we patch in order not to confuse our natural pattern matcher (I didn't manage to get something useful from meld there)

Reply:
PS: I wonder if it's worthwhile to try and replace this patch, the one in https://github.com/OCA/server-tools/blob/17.0/upgrade_analysis/odoo_patch/odoo/tools/convert.py and https://github.com/OCA/server-tools/blob/17.0/upgrade_analysis/odoo_patch/odoo/models.py with a patch on https://github.com/OCA/OCB/blob/17.0/odoo/addons/base/models/ir_model.py#L2334?

This we can even do without copying the code and avoid this class of problem altogether. Will test this in a couple of hours when my current test run for what I did above is done.

Reply:
Thanks for checking, and I like your suggestion to patching the xmlid methods themselves.

Reply:
my first test of this failed and currently I can't dive into this, so let's go with what we have, that generates correct files. squashed the fixup commit